### PR TITLE
Add Schema Type to AWS::Serverless::Lambda

### PR DESCRIPTION
**Pull request by GavinZZ**
### Issue #, if available

#1133 

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [unit tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions) using:
    - [ ] Correct values
    - [ ] Bad/wrong values (None, empty, wrong type, length, etc.)
    - [ ] [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html)
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)
- [ ] Update documentation
- [ ] Verify transformed template deploys and application functions as expected
- [ ] Do these changes include any template validations?
    - [ ] Did the newly validated properties support intrinsics prior to adding the validations? (If unsure, please review [Intrinsic Functions](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html) before proceeding).
        - [ ] Does the pull request ensure that intrinsics remain functional with the new validations?

### Examples?

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/translator/input/api_with_method_aws_iam_auth.yaml [hunk 1]
```diff
@@ -45,21 +45,21 @@ Resources:
...
+              InvokeRole: !Sub arn:${AWS::Partition}:iam::123:role/AUTH_AWS_IAM
+        MyApiWithAwsIamAuthAndDefaultInvokeRoleAnyMethod:
+          Type: Api
+          Properties:
+            RestApiId: !Ref MyApiWithoutAuth
+            Path: /any/three
+            Method: any
+            Auth:
+              Authorizer: AWS_IAM
+              InvokeRole: CALLER_CREDENTIALS
```
Comment (aahung) [inline tests/translator/input/api_with_method_aws_iam_auth.yaml:65]: unbelievable!

File: tests/translator/input/eventbridgerule_with_retry_policy.yaml [hunk 2]
```diff
@@ -10,8 +10,8 @@ Resources:
           Type: Schedule
           Properties:
             Schedule: rate(1 minute)
-          RetryPolicy:
-            MaximumRetryAttempts: 3
+            RetryPolicy:
+              MaximumRetryAttempts: 3
```
Comment (aahung) [inline tests/translator/input/eventbridgerule_with_retry_policy.yaml:14]: :o

File: tests/translator/input/function_with_deployment_preference_from_parameters.yaml [hunk 3]
```diff
@@ -40,6 +40,7 @@ Resources:
       CodeUri: s3://my-bucket/mySanityTestFunction.zip
       DeploymentPreference:
         Enabled: !Ref MyLowerFalseParameter
+        Type: Linear10PercentEvery1Minute
```
Comment (aahung) [inline tests/translator/input/function_with_deployment_preference_from_parameters.yaml:44]: what are these changes?

File: tests/translator/input/function_with_deployment_preference_all_parameters.yaml [hunk 4]
```diff
@@ -38,6 +39,7 @@ Resources:
       CodeUri: s3://my-bucket/myValidationTestFunction.zip
       DeploymentPreference:
         Enabled: false
+        Type: Linear10PercentEvery1Minute
```
Comment (hoffa) [inline tests/translator/input/function_with_deployment_preference_all_parameters.yaml:42]: Why are you changing these? We should be first of all testing against the status quo, with no changes.

File: tests/translator/input/function_with_deployment_preference_from_parameters.yaml (hunk 3 above)
Comment (GavinZZ) [reply tests/translator/input/function_with_deployment_preference_from_parameters.yaml:44]: It's a required field, but it's missing. Json schema won't let it pass.

File: samtranslator/schema/schema.py [hunk 5]
```diff
@@ -45,42 +45,84 @@ class AwsServerlessConnector(BaseModel):
...
+    Hooks: Optional[FunctionDeploymentPreferenceHooks]
+    PassthroughCondition: Optional[Union[bool, SamIntrinsic]]
+    Role: Optional[Union[str, SamIntrinsic]]
+    TriggerConfigurations: Optional[PassThrough]
+    Type: Union[str, SamIntrinsic]
+
+
+class FunctionDeadLetterQueue(BaseModel):
+    TargetArn: str
+    Type: Literal["SNS", "SQS"]
```
Comment (hoffa) [inline samtranslator/schema/schema.py:71]: Let's separate these so the `Type` is coupled to whatever properties it has. Otherwise makes it a pain to expand later.

File: tests/translator/input/function_with_deployment_preference_all_parameters.yaml (hunk 4 above)
Comment (GavinZZ) [reply tests/translator/input/function_with_deployment_preference_all_parameters.yaml:42]: It's a required field, but missing from the template. This results json schema check failure. 
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-deploymentpreference.html#sam-function-deploymentpreference-type

File: samtranslator/schema/schema.py [hunk 6]
```diff
@@ -45,42 +45,84 @@ class AwsServerlessConnector(BaseModel):
...
+
+
+class FunctionEventInvokeConfiguration(BaseModel):
+    DestinationConfig: Optional[PassThrough]
+    MaximumEventAgeInSeconds: Optional[int]
+    MaximumRetryAttempts: Optional[int]
+
+
+class FunctionEvent(BaseModel):
+    Properties: Any
```
Comment (hoffa) [inline samtranslator/schema/schema.py:81]: Missing types. Also same as comment above; should couple both `Properties` and `Type` so that each model only has a single `Type`.

File: tests/translator/input/function_with_deployment_preference_all_parameters.yaml (hunk 4 above)
Comment (hoffa) [reply tests/translator/input/function_with_deployment_preference_all_parameters.yaml:42]: Let's skip it. Let's not change any templates currently. There is no formal schema at the moment. First step is to document what our tests consider as valid, as that is the effective source of truth (since that's what deal with the logic). We should document the discrepancies and later figure out whether docs need to change, or tests need to change.

File: samtranslator/schema/schema.py (hunk 5 above)
Comment (hoffa) [reply samtranslator/schema/schema.py:71]: Ignore, this one is fine. Still applies to `Events`.

File: Makefile [hunk 7]
```diff
@@ -34,8 +34,6 @@ lint:
 	mypy --strict samtranslator bin
 	# Linter performs static analysis to catch latent bugs
 	pylint --rcfile .pylintrc samtranslator
-	# Ensure templates adhere to JSON schema
-	bin/validate.sh
```
Comment (hoffa) [inline Makefile:38]: Should not be removed. We need to ensure all templates going forward adhere to the schema.

File: samtranslator/schema/schema.py (hunk 6 above)
Comment (GavinZZ) [reply samtranslator/schema/schema.py:81]: Will do this in a separate PR. It's a huge change.

File: tests/translator/output/aws-us-gov/api_with_method_aws_iam_auth.json [hunk 8]
```diff
@@ -100,9 +140,9 @@
       },
       "Type": "AWS::ApiGateway::RestApi"
     },
-    "MyApiWithoutAuthDeployment60e94c7d4c": {
+    "MyApiWithoutAuthDeployment54f5d55d46": {
```
Comment (hoffa) [inline tests/translator/output/aws-us-gov/api_with_method_aws_iam_auth.json:143]: Something's off; this shouldn't change or it'll cause redeployments.

File: tests/translator/input/api_with_method_aws_iam_auth.yaml [hunk 9]
```diff
@@ -27,7 +27,7 @@ Resources:
             Method: post
             Auth:
               Authorizer: AWS_IAM
-              InvokeRole: rn:aws:iam::123:role/AUTH_AWS_IAM
+              InvokeRole: !Sub arn:${AWS::Partition}:iam::123:role/AUTH_AWS_IAM
```
Comment (hoffa) [inline tests/translator/input/api_with_method_aws_iam_auth.yaml:30]: Why this change?

File: bin/validate.sh [hunk 10]
```diff
@@ -3,6 +3,6 @@ set -euxo pipefail
 
 # TODO: Enable the valid templates
 # TODO: Switch to way-faster Python script
-find tests/translator/input -type f -name '*.yaml' | grep -v error_ | grep -v unsupported_resources | grep -v resource_with_invalid_type | shuf | while read -r template; do
+find tests/translator/input -type f -name '*.yaml' | grep -v error_ | grep -v null_events | grep -v unsupported_resources | grep -v resource_with_invalid_type |  while read -r template; do
```
Comment (hoffa) [inline bin/validate.sh:6]: Was `null_events` supposed to skip two tests? If yes, clearer:

```suggestion
find tests/translator/input -type f -name '*.yaml' | grep -v error_ | grep -v function_with_null_events | grep -v state_machine_with_null_events | grep -v unsupported_resources | grep -v resource_with_invalid_type |  while read -r template; do
```

Comment (GavinZZ) [reply bin/validate.sh:6]: Yes, it's supposed to skip two tests.

File: tests/translator/input/api_with_method_aws_iam_auth.yaml (hunk 9 above)
Comment (GavinZZ) [reply tests/translator/input/api_with_method_aws_iam_auth.yaml:30]: The arn was incorrect `rn:aws:iam...`.

Comment (hoffa) [reply tests/translator/input/api_with_method_aws_iam_auth.yaml:30]: Okay, so I'm assuming it transformed correctly. Transform tests should ideally never change, so we can skip making these changes as well (maybe make a note for later cleanup) for easier review (e.g. I'm not sure whether the logical ID change is due to this or the indentation change).

File: samtranslator/schema/schema.py [hunk 11]
```diff
@@ -52,42 +51,86 @@ class AwsServerlessConnector(BaseModel):
...
+    AutoPublishAlias: Optional[Union[str, SamIntrinsic]]
+    AutoPublishCodeSha256: Optional[Union[str, SamIntrinsic]]
+    CodeSigningConfigArn: Optional[Union[str, SamIntrinsic]]
+    CodeUri: Optional[Union[str, FunctionCodeUri]]
+    DeadLetterQueue: Optional[Union[SamIntrinsic, FunctionDeadLetterQueue]]
+    DeploymentPreference: Optional[FunctionDeploymentPreference]
+    Description: Optional[PassThrough]
+    Environment: Optional[PassThrough]
+    EphemeralStorage: Optional[PassThrough]
+    EventInvokeConfig: Optional[FunctionEventInvokeConfiguration]
```
Comment (hoffa) [inline samtranslator/schema/schema.py:110]: nit: we'll have a lot of names, so keeping it consistent throughout will make it easier to reason (i.e. use `Config` if that's what the property uses)

File: tests/translator/input/api_with_method_aws_iam_auth.yaml (hunk 9 above)
Comment (GavinZZ) [reply tests/translator/input/api_with_method_aws_iam_auth.yaml:30]: The transform is correct, and the id change is due to indentation change I believe. It's when I make the indentation change, I found the arn seems off. And there are two occurrence of wrong arn format.

**Review by hoffa — APPROVED** [review]

**Review by aahung — APPROVED** [review]
